File: src/main/java/com/silverboxsoft/dynamodbtool/classes/DynamoDbResult.java
package com.silverboxsoft.dynamodbtool.classes;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import com.silverboxsoft.dynamodbtool.utils.DynamoDbUtils;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.TableDescription;

public class DynamoDbResult {

	private List<DynamoDbColumn> columnList = new ArrayList<>();
	private Map<String, Integer> colNameIndex = new HashMap<>();
	private List<DynamoDbViewRecord> resItems = new ArrayList<>();
	private List<Map<String, AttributeValue>> rawResItems = new ArrayList<>();

	public DynamoDbResult(List<Map<String, AttributeValue>> items, TableDescription tableInfo) {
		prepareKeyInfo(tableInfo);
		initilize(items);
	}

	private void prepareKeyInfo(TableDescription tableInfo) {
		columnList = DynamoDbUtils.getSortedDynamoDbColumnList(tableInfo);
		for (int idx = 0; idx < columnList.size(); idx++) {
			colNameIndex.put(columnList.get(idx).getColumnName(), idx);
		}
	}

	private void initilize(List<Map<String, AttributeValue>> items) {
		for (Map<String, AttributeValue> resItem : items) {
			addRecord(resItem);
		}
	}

	public ObservableList<String> prepareOneTableRecord(Map<String, AttributeValue> resItem) {
		ObservableList<String> record = FXCollections.observableArrayList();

		// pick up data which attribute name is set.
		for (int colIdx = 0; colIdx < columnList.size(); colIdx++) {
			DynamoDbColumn dbCol = getDynamoDbColumn(colIdx);
			String columnName = dbCol.getColumnName();
			record.add(DynamoDbUtils.getAttrString(resItem.get(columnName)));
		}

		// pick up the data which attribute name is come yet.
		Set<String> unsetKeyNameSet = resItem.keySet().stream().collect(Collectors.toSet());
		unsetKeyNameSet.removeAll(colNameIndex.keySet());
		for (String newColName : unsetKeyNameSet) {
			AttributeValue attrVal = resItem.get(newColName);
			DynamoDbColumn dbCol = new DynamoDbColumn();
			dbCol.setColumnName(newColName);
			dbCol.setColumnType(DynamoDbUtils.getDynamoDbColumnType(attrVal));
			fillNewColumn();
			record.add(DynamoDbUtils.getAttrString(attrVal));
			colNameIndex.put(newColName, columnList.size());
			columnList.add(dbCol);
		}
		return record;
	}

	private void fillNewColumn() {
		this.resItems.stream().forEach(rec -> rec.getData().add(DynamoDbUtils.NO_VALSTR));
	}

	public int getColumnCount() {
		return columnList.size();
	}

	public int getRecordCount() {
		return resItems.size();
	}

	public List<DynamoDbViewRecord> getResultItems() {
		return resItems;
	}

	public List<Map<String, AttributeValue>> getRawResItems() {
		return rawResItems;
	}

	public DynamoDbColumn getDynamoDbColumn(int index) {
		return columnList.get(index);
	}

	public DynamoDbViewRecord addRecord(Map<String, AttributeValue> newRec) {
		this.rawResItems.add(newRec);

		ObservableList<String> data = prepareOneTableRecord(newRec);
		DynamoDbViewRecord record = DynamoDbViewRecord.builder().index(this.resItems.size()).data(data)
				.build();
		this.resItems.add(record);
		return record;
	}

	public DynamoDbViewRecord updateRecord(int rowIndex, Map<String, AttributeValue> newRec) {
		rawResItems.set(rowIndex, newRec);
		ObservableList<String> data = prepareOneTableRecord(newRec);
		DynamoDbViewRecord record = resItems.get(rowIndex);
		record.setData(data);
		resItems.set(rowIndex, record);
		return record;
	}

	public void removeRecord(int rowIndex) {
		rawResItems.remove(rowIndex);
	}

	public Integer getColumnIndexByName(String colName) {
		if (colNameIndex.containsKey(colName)) {
			return colNameIndex.get(colName);
		}
		return -1;
	}
}

File: src/main/java/com/silverboxsoft/dynamodbtool/classes/DynamoDbResult.kt
package com.silverboxsoft.dynamodbtool.classes

import java.util.HashMap
import com.silverboxsoft.dynamodbtool.utils.DynamoDbUtils
import java.util.stream.Collectors
import javafx.collections.ObservableList
import javafx.collections.FXCollections

import software.amazon.awssdk.services.dynamodb.model.*
import java.util.ArrayList

class DynamoDbResult(items: List<Map<String, AttributeValue>>, tableInfo: TableDescription) {
    private var columnList: MutableList<DynamoDbColumn> = ArrayList()
    private val colNameIndex: MutableMap<String, Int> = HashMap()
    private val resItems: MutableList<DynamoDbViewRecord> = ArrayList()
    private val rawResItems: MutableList<Map<String, AttributeValue>> = ArrayList()

    private fun prepareKeyInfo(tableInfo: TableDescription) {
        columnList = DynamoDbUtils.Companion.getSortedDynamoDbColumnList(tableInfo)
        for (idx in columnList.indices) {
            colNameIndex[columnList[idx].columnName] = idx
        }
    }

    private fun initilize(items: List<Map<String, AttributeValue>>) {
        for (resItem in items) {
            addRecord(resItem)
        }
    }

    fun prepareOneTableRecord(resItem: Map<String, AttributeValue>): ObservableList<String> {
        val record = FXCollections.observableArrayList<String>()

        // pick up data which attribute name is set.
        for (colIdx in columnList.indices) {
            val dbCol = getDynamoDbColumn(colIdx)
            val columnName = dbCol.columnName
            record.add(DynamoDbUtils.Companion.getAttrString(resItem[columnName]))
        }

        // pick up the data which attribute name is come yet.
        val unsetKeyNameSet = resItem.keys.stream().collect(Collectors.toSet())
        unsetKeyNameSet.removeAll(colNameIndex.keys)
        for (newColName in unsetKeyNameSet) {
            val attrVal = resItem[newColName]
            val dbCol = DynamoDbColumn(newColName, DynamoDbUtils.Companion.getDynamoDbColumnType(attrVal))
            fillNewColumn()
            record.add(DynamoDbUtils.Companion.getAttrString(attrVal))
            colNameIndex[newColName] = columnList.size
            columnList.add(dbCol)
        }
        return record
    }

    private fun fillNewColumn() {
        resItems.stream().forEach { rec: DynamoDbViewRecord -> rec.getData().plus(DynamoDbUtils.Companion.NO_VALSTR) }
    }

    val columnCount: Int
        get() = columnList.size
    val recordCount: Int
        get() = resItems.size
    val resultItems: List<DynamoDbViewRecord>
        get() = resItems

    fun getRawResItems(): List<Map<String, AttributeValue>> {
        return rawResItems
    }

    fun getDynamoDbColumn(index: Int): DynamoDbColumn {
        return columnList[index]
    }

    fun addRecord(newRec: Map<String, AttributeValue>): DynamoDbViewRecord {
        rawResItems.add(newRec)
        val data = prepareOneTableRecord(newRec)
        val record: DynamoDbViewRecord = DynamoDbViewRecord(resItems.size, data)
        resItems.add(record)
        return record
    }

    fun updateRecord(rowIndex: Int, newRec: Map<String, AttributeValue>): DynamoDbViewRecord {
        rawResItems[rowIndex] = newRec
        val data = prepareOneTableRecord(newRec)
        val record = resItems[rowIndex]
        record.setData(data)
        resItems[rowIndex] = record
        return record
    }

    fun removeRecord(rowIndex: Int) {
        rawResItems.removeAt(rowIndex)
    }

    fun getColumnIndexByName(colName: String): Int {
        return colNameIndex.getOrDefault(colName, -1)
    }

    init {
        prepareKeyInfo(tableInfo)
        initilize(items)
    }
}